src/storage-client/src/source/persist_source.rs

```diff
@@ -134,465 +126,149 @@ where
-        // through, including the initial downgrade of the shard upper to the
-        // `as_of`.
-        yield (Vec::new(), as_of_stream.clone(), 0);
-
-        // Always poll drop_rx first. drop_rx returns only if drop_tx has been dropped.
-        // In this case, we deliberately do not continue on the subscribe. but terminate this task
-        let subscription = tokio::select! {
-            biased;
-            _ = drop_rx => None,
-            x = read.subscribe(as_of_stream.clone()) => Some(x),
-        };
-
-        // This is a moment where we may have let persist compact if our token
-        // has been dropped, but if we still hold it we should be good to go.
-        if weak_handle_token.upgrade().is_none() {
-            return;
-        }
-
-        // We get here only when the token is still present, hence the subscription select
-        // must have returned from the subscribe and hence we can be sure that subscripbtion
-        // is Some.
-        let subscription = subscription.unwrap();
-
-        let mut subscription = subscription.unwrap_or_else(|e| {
-            panic!(
-                "{source_id}: {} cannot serve requested as_of {:?}: {:?}",
-                data_shard, as_of_stream, e
-            )
-        });
-
-        let mut done = false;
-        while !done {
-            while let Ok(leased_part) = consumed_part_rx.try_recv() {
-                subscription
-                    .return_leased_part(subscription.leased_part_from_exchangeable(leased_part));
-            }
-
-            let (parts, progress) = subscription.next().await;
-            // If `until.less_equal(progress)`, it means that all subsequent batches will
-            // contain only times greater or equal to `until`, which means they can be dropped
-            // in their entirety. The current batch must be emitted, but we can stop afterwards.
-            if PartialOrder::less_equal(&until_clone, &progress) {
-                done = true;
-            }
-
-            let parts_size_bytes = parts.iter().map(|x| x.encoded_size_bytes()).sum();
-            yield (parts, progress, parts_size_bytes);
-        }
-
-        // Rather than simply end, we spawn a task that can continue to return leases.
-        // This keeps the `ReadHandle` alive until we have confirmed each lease as read.
-        mz_ore::task::spawn(|| "LeaseReturner", async move {
-            while let Some(leased_part) = consumed_part_rx.recv().await {
-                subscription
-                    .return_leased_part(subscription.leased_part_from_exchangeable(leased_part));
-            }
-        });
-    });
-
-    let mut pinned_stream = Box::pin(async_stream);
-
-    let mut builder_dist = AsyncOperatorBuilder::new(
-        format!("persist_source {}: part distribution", source_id),
-        scope.clone(),
+    let name = source_id.to_string();
+    let (fetched, token) = shard_source(
```

Comment:
don't we eventually have to push the mfp all the way into the shard_source?

Reply:
yeah, certainly has the potential to be the most controversial part of this, but I'd like to avoid MapFilterProject being the interface to persist (in the same way we do for Row). TBD a bunch of work for what this actual abstraction looks like, but I'm imagining something morally equivalent to the way Codec abstracts things now (but with necessarily more detail)